### PR TITLE
Add `rediss` (Redis over SSL) protocol to `RedisDsn`

### LINKED ISSUE #1877: RedisDsn canot enable redis url with password and ssl schema
`redis://:password@redishost` & `rediss://redis` are valid in redis.py, but cause ValidationError in RedisDsn. 

```
In [7]: import redis

In [8]: r = redis.StrictRedis.from_url('rediss://redis')

In [9]: r = redis.StrictRedis.from_url('redis://:123@redis')

In [10]:
```

* type error list below
```
In [3]: from pydantic import BaseModel, RedisDsn

In [4]: class RedisURLModel(BaseModel):
   ...:     url: RedisDsn = None
   ...:

In [5]: c = RedisURLModel(url='redis://:myredispwd@redis')
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
<ipython-input-5-23cad9c7fe6b> in <module>
----> 1 c = RedisURLModel(url='redis://:myredispwd@redis')

/usr/local/lib/python3.7/site-packages/pydantic/main.cpython-37m-x86_64-linux-gnu.so in pydantic.main.BaseModel.__init__()

ValidationError: 1 validation error for RedisURLModel
url
  URL host invalid (type=value_error.url.host)

In [6]: c = RedisURLModel(url='rediss://redis')
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
<ipython-input-6-39558dddb3bc> in <module>
----> 1 c = RedisURLModel(url='rediss://redis')

/usr/local/lib/python3.7/site-packages/pydantic/main.cpython-37m-x86_64-linux-gnu.so in pydantic.main.BaseModel.__init__()

ValidationError: 1 validation error for RedisURLModel
url
  URL scheme not permitted (type=value_error.url.scheme; allowed_schemes={'redis'})
```

### PR DESCRIPTION
**Pull request by TrDex**
## Change Summary

Added `rediss` to `RedisDsn.allowed_schemes`.

## Related issue number

This PR closes #1877 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/1911-TrDex.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: changes/1911-TrDex.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Add `rediss` (Redis over SSL) protocol to `RedisDsn
```
Comment (PrettyWood) [inline changes/1911-TrDex.md:1]:
```suggestion
Add `rediss` (Redis over SSL) protocol to `RedisDsn`
```

**Review by PrettyWood — COMMENTED** [review]
Small typo in changelog. Otherwise LGTM !

**Comment by PrettyWood** [comment]
@TrDex I think you should link this issue #1877 as your PR adds `rediss`.
You could also go a bit further by changing [here](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/networks.py#L57) `?P<user>[^\s:/]+` into `?P<user>[^\s:/]*`.
This way you could add both examples `'rediss://redis'` and `'redis://:123@redis'` and fully solve the issue. WDYT ?

**Comment by TrDex** [comment]
> @TrDex I reckon you should link this issue #1877 as your PR adds `rediss`.

Thanks a lot @PrettyWood ! I updated the PR comment (hope it works that way, that's one of my first PRs, lol)

> You could also go a bit further by changing [here](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/networks.py?rgh-link-date=2020-09-09T23%3A11%3A40Z#L57) `?P<user>[^\s:/]+` into `?P<user>[^\s:/]*`.
> This way you could add both examples `'rediss://redis'` and `'redis://:123@redis'` in your tests and fully solve the issue. WDYT ?

Okay, I'll update that too, and make corresponding tests as well

File: tests/requirements.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-black==19.10b0
+black==20.8b1
```
Comment (samuelcolvin) [inline tests/requirements.txt:1]: best not to change requirements in this PR.

File: pydantic/utils.py [hunk 3]
```diff
@@ -241,7 +241,8 @@ def unique_list(input_list: Union[List[T], Tuple[T, ...]]) -> List[T]:
 
 
 def update_normalized_all(
-    item: Union['AbstractSetIntStr', 'MappingIntStrAny'], all_items: Union['AbstractSetIntStr', 'MappingIntStrAny'],
+    item: Union['AbstractSetIntStr', 'MappingIntStrAny'],
```
Comment (samuelcolvin) [inline pydantic/utils.py:244]: this doesn't need changing.

File: pydantic/networks.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ def url_regex() -> Pattern[str]:
     if _url_regex_cache is None:
         _url_regex_cache = re.compile(
             r'(?:(?P<scheme>[a-z][a-z0-9+\-.]+)://)?'  # scheme https://tools.ietf.org/html/rfc3986#appendix-A
-            r'(?:(?P<user>[^\s:/]+)(?::(?P<password>[^\s/]*))?@)?'  # user info
+            r'(?:(?P<user>[^\s:/]*)(?::(?P<password>[^\s/]*))?@)?'  # user info
```
Comment (samuelcolvin) [inline pydantic/networks.py:57]: this is quite a significant change - allow passing without user.

I think it's a good idea but it needs:
* to be noted in documentation
* to have it's own item in change log
* to have tests covering this case.

File: pydantic/utils.py (hunk 3 above)
Comment (TrDex) [reply pydantic/utils.py]: PEP8 says that a line has to be less than 79 characters

File: pydantic/networks.py (hunk 4 above)
Comment (samuelcolvin) [reply pydantic/networks.py:57]: these items still need resolving.

**Comment by nuno-andre** [comment]
Additionally:
- The path, if any, should only be numeric (without leading zeros) and with a single segment.
- All fields after the scheme are optional. `hostname` defaults to `localhost`, `port` to `6379` and `path` to `0`.

https://www.iana.org/assignments/uri-schemes/prov/redis

File: pydantic/networks.py (hunk 4 above)
Comment (samuelcolvin) [reply pydantic/networks.py:56]: @TrDex this can't move forward until you act upon this comment.

File: pydantic/networks.py [hunk 5]
```diff
@@ -279,7 +289,17 @@ class PostgresDsn(AnyUrl):
...
+    allowed_schemes = {'redis', 'rediss'}
+
+    @classmethod
+    def validate_parts(cls, parts: Dict[str, Optional[str]]) -> Dict[str, Optional[str]]:
+        defaults = {
+            'domain': 'localhost' if not (parts['ipv4'] or parts['ipv6']) else None,
+            'port': '6379',
+            'path': '/0',
+        }
+        parts = {k: v if v else defaults.get(k) for k, v in parts.items()}
```
Comment (samuelcolvin) [inline pydantic/networks.py:301]: probably marginally more efficient and clear to do something like

```py
for key, value in defaults.items():
    if not parts[key]:
        parts[key] = value
```
or similar.

**Review by samuelcolvin — COMMENTED** [review]
Also please fix linting.

**Comment by TrDex** [comment]
@samuelcolvin I worked on your comments about allowing password without user.

- [x] It's noted in the documentation.
- [x] It has its own item in the changelog.
- [x] It has tests covering this case.

File: tests/test_networks.py [hunk 6]
```diff
@@ -321,21 +328,35 @@ def test_redis_dsns():
     class Model(BaseModel):
         a: RedisDsn
 
-    m = Model(a='redis://user:pass@localhost:5432/app')
-    assert m.a == 'redis://user:pass@localhost:5432/app'
+    m = Model(a='redis://user:pass@localhost:6379/app')
```
Comment (PrettyWood) [inline tests/test_networks.py:331]: I reckon it's better to keep `5432` to first try with a different port than the default one

Comment (daviskirk) [reply tests/test_networks.py:331]: Perhaps 5432 is a bad choice as an "non-default" port as it's quite well known as the postgres port and people might be confused (or think it's a typo).

Comment (nuno-andre) [reply tests/test_networks.py:331]: @daviskirk  I agree. A dynamic port (49152-65535) may be a better choice.

**Pull request merged**